tests/mir-opt/inline/unwrap_unchecked.unwrap_unchecked.PreCodegen.after.mir

```diff
@@ -0,0 +1,41 @@
+// MIR for `unwrap_unchecked` after PreCodegen
+
+fn unwrap_unchecked(_1: Option<T>) -> T {
+    debug slf => _1;                     // in scope 0 at $DIR/unwrap_unchecked.rs:+0:35: +0:38
+    let mut _0: T;                       // return place in scope 0 at $DIR/unwrap_unchecked.rs:+0:54: +0:55
+    scope 1 (inlined #[track_caller] Option::<T>::unwrap_unchecked) { // at $DIR/unwrap_unchecked.rs:7:9: 7:27
+        debug self => _1;                // in scope 1 at $SRC_DIR/core/src/option.rs:LL:COL
+        let mut _2: &std::option::Option<T>; // in scope 1 at $SRC_DIR/core/src/option.rs:LL:COL
+        let mut _3: isize;               // in scope 1 at $SRC_DIR/core/src/option.rs:LL:COL
+        scope 2 {
+            debug val => _0;             // in scope 2 at $SRC_DIR/core/src/option.rs:LL:COL
+        }
+        scope 3 {
+            scope 5 (inlined unreachable_unchecked) { // at $SRC_DIR/core/src/option.rs:LL:COL
+                scope 6 {
+                    scope 7 (inlined unreachable_unchecked::runtime) { // at $SRC_DIR/core/src/intrinsics.rs:LL:COL
+                    }
+                }
+            }
+        }
+        scope 4 (inlined Option::<T>::is_some) { // at $SRC_DIR/core/src/option.rs:LL:COL
+            debug self => _2;            // in scope 4 at $SRC_DIR/core/src/option.rs:LL:COL
+        }
+    }
+
+    bb0: {
+        StorageLive(_2);                 // scope 0 at $DIR/unwrap_unchecked.rs:+1:9: +1:27
+        _3 = discriminant(_1);           // scope 1 at $SRC_DIR/core/src/option.rs:LL:COL
+        switchInt(move _3) -> [1: bb2, otherwise: bb1]; // scope 1 at $SRC_DIR/core/src/option.rs:LL:COL
```

Comment:
Idea for a follow-up: make this a `goto bb2` as `bb1` in unreachable.

Reply:
That's a good experiment, but consider this code:
```rust
unsafe fn assert_nonzero(x: u8) -> u8 {
    match x {
        0 => core::hint::unreachble_unchecked(),
        _ => x,
    }
}
```
I think you're suggesting that we optimize this to just return the argument, which would be valid, but also means that code after this now needs to consider the case where `x == 0`.

Reply:
Agreed with saethlin here -- better to leave the `unreachable` there for LLVM to see, since it's better at back-propagating the "`x` can't be zero" than we can reasonably be in MIR.

Though I guess it could emit an [`Assume`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.NonDivergingIntrinsic.html#variant.Assume) + `goto`?  Dunno if that's a good idea, though, since assumes can actually interrupt other optimizations in LLVM sometimes.  Might still be best to leave the consumption of `unreachable` to the backend.